couchcopy

```diff
@@ -59,41 +59,69 @@ async def subprocess(*args, shell=False, input=None):
-    await subprocess('rsync', '--del', '-av', hostname + path + '/_dbs.couch',
-                     tmp_path + '/data')
-    await subprocess('rsync', '--del', '--ignore-missing-args',
-                     '-av', hostname + path + '/shards', tmp_path + '/data')
-
-    print('[tar...]')
-    await subprocess('tar', '-I', 'pigz', '-cf', output, '-C', tmp_path,
-                     'metadata.yaml', 'data')
-    # For debugging purposes:
-    # if tmp_dir:
-    #     shutil.rmtree('/tmp/couchcopy', True)
-    #     shutil.copytree(tmp_dir.name, '/tmp/couchcopy')
+async def backup(hostname, path, output, nodes_names=None):
+    metadata = yaml.dump({
+        'backup': {
+            'date': datetime.now().isoformat(),
+            'source': {'nodes-names': nodes_names or []},
+        },
+        'couchcopy': {'version': __version__},
+    }, default_flow_style=False).encode()
+
+    print('[tar+ssh+gzip...]')
+
+    ssh_if_needed = ('ssh', hostname, '--') if hostname != 'localhost' else ()
+    tempdir = await subprocess(*ssh_if_needed,
+                               'mktemp', '-d', '-t', 'couchcopy-XXXXX')
+
+    try:
+        tempdir = tempdir.strip()
+        await subprocess(*ssh_if_needed,
+                         'dd', f'of={tempdir}/metadata.yaml', 'status=none',
+                         input=metadata)
+
+        with open(output, 'wb') as archive:
+            pipe_out, pipe_in = os.pipe()
+            p1 = await asyncio.create_subprocess_exec(
+                *ssh_if_needed,
+                'tar',
+                '-C', tempdir, '-c', 'metadata.yaml',
+                '-C', os.path.dirname(path),
+                '-c', os.path.basename(path) + '/.shards',
+                os.path.basename(path) + '/_dbs.couch',
+                os.path.basename(path) + '/shards',
+                '--transform', f's/^{re.escape(os.path.basename(path))}/data/',
+                stdout=pipe_in, stderr=asyncio.subprocess.PIPE,
+                env={'LANG': 'C'})
+            p2 = await asyncio.create_subprocess_exec(
+                'pigz', stdin=pipe_out, stdout=archive,
+                stderr=asyncio.subprocess.PIPE)
+            os.close(pipe_in)
+            os.close(pipe_out)
+            _, stderr1, returncode1 = *await p1.communicate(), p1.returncode
+            _, stderr2, returncode2 = *await p2.communicate(), p2.returncode
+
+            # tar can exit with return code 1 if files are modified on disk
+            # during the archive creation. This is normal behavior when the
+            # CouchDB instance is running, so let's filter out these errors and
+            # fail only on other ones.
+            if returncode1 == 1:
+                stderr1 = '\n'.join(
+                    l for l in stderr1.decode().splitlines()
+                    if not l.endswith('file changed as we read it')).encode()
+                if not stderr1:
+                    returncode1 = 0
```

Comment:
To test this error handling, here is a simple Bash script that continuously modifies files inside the source CouchDB directory:

```bash
#!/bin/bash
set -eu
shards="…/couchdb/shards/00000000-7fffffff"
file=
for i in $(seq 1000000); do
  [ -n "$file" ] && rm $file; file=$shards/aaa_file_$i
  dd if=/dev/urandom of=$file bs=1024 count=1024 status=none
done
rm $file
```

Reply:
Just by curiosity, if a file is modified while `tar` reads it, which version of the file ends up in the archive: the original one or the new one? And how does `rsync` behave in the same case?

Reply:
Very good question. I didn't find anything about in on the internet, so I tested in on my laptop. The answer seems to be "the new one".
##### Test with `tar`

```bash
# Prepare 1000 files filled with 'A':
$ ./write-1000-files-with-byte.sh A
# Create archive and fill files with 'B' at the same time:
$ tar -f /tmp/backup.tar.gz -C /tmp -c test & ./write-1000-files-with-byte.sh B; wait
tar: test/1MB_file_968.txt : fichier modifié pendant sa lecture
# Check the content of errored file:
$ tar -axf /tmp/backup.tar.gz test/1MB_file_968.txt -O | head -c 10
BBBBBBBBBB%
```
↑ I repeated the test 6 times, I always got the "new" version

##### Test with `rsync`

```bash
# Prepare 1000 files filled with 'A':
$ ./write-1000-files-with-byte.sh A
# Rsync data and fill files with 'B' at the same time:
$ rm -rf /tmp/rsync; rsync -a /tmp/test /tmp/rsync/ & ./write-1000-files-with-byte.sh B; wait
rsync: [sender] read errors mapping "/tmp/test/1MB_file_115.txt": No data available (61)
$ head -c 10 /tmp/rsync/test/1MB_file_115.txt
BBBBBBBBBB%    
```
↑ I repeated the test 4 times, I always got the "new" version

File `./write-1000-files-with-byte.sh`:
```bash
for i in $(seq 1000); do
  head -c 1048576 /dev/zero | tr "\0" "$1" >/tmp/test/1MB_file_$i.txt &
done
wait
```

What do you think?

Reply:
Interesting, thanks for investigating this! Nice to see that `tar` and `rsync` behaves in the same way :+1: